protocols/gossipsub/src/protocol.rs

```diff
@@ -0,0 +1,385 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::rpc_proto;
+use byteorder::{BigEndian, ByteOrder};
+use bytes::BytesMut;
+use futures::{future, stream, Future, Stream};
+use libp2p_core::{InboundUpgrade, OutboundUpgrade, PeerId, UpgradeInfo};
+use libp2p_floodsub::TopicHash;
+use protobuf::Message as ProtobufMessage;
+use std::{io, iter};
+use tokio_codec::{Decoder, FramedRead};
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint::codec;
+
+/// Implementation of the `ConnectionUpgrade` for the Gossipsub protocol.
+#[derive(Debug, Clone, Default)]
+pub struct ProtocolConfig {}
+
+impl ProtocolConfig {
+    /// Builds a new `ProtocolConfig`.
+    #[inline]
+    pub fn new() -> ProtocolConfig {
+        ProtocolConfig {}
+    }
+}
+
+impl UpgradeInfo for ProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    #[inline]
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/meshsub/1.0.0")
```

Comment:
To expand on #880, the idea is that you can return multiple protocol names here. You can return both `/meshsub/1.0.0` and `/floodsub/1.0.0`. Then `upgrade_inbound` and `upgrade_outbound` receive the protocol name and can generate a different output. This way you can support both protocols at the same time.

In practice, it is probably easier either:
- To have two different implementations of `ProtocolsHandler` (one for gossipsub, one for floodsub) and merge then with `ProtocolsHandler::select()`.
- To have two different implementations of `UpgradeInfo`/`InboundUpgrade`/`OutboundUpgrade` and merge them with `or_upgrade`.

Reply:
Hey @tomaka, 
Thanks for this. The spec seems to want to be backwards compatible with floodsub. It seems to me that I could provide two implementations for `upgrade_inbound` and `upgrade_outbound`, however this would only allow me to modify the input/output data to nodes for each protocol. 

I need to perform "behaviour-level" logic. For example, sending a message to topic `X`, should send to all peers that implement the gossipsub protocol that are in `mesh` and also to all floodsub peers.  I can't see an easy way of implementing this kind of logic at the `ProtocolsHandler` level (with multiple `upgrade_inbound`/`upgrade_outbound` functions). 

The `floodsub` behaviour in `gossipsub` just mimics the typical `floodsub` behaviour. It seems to me the best approach would be to strip out all floodsub logic from this implementation, and if users wish to be backwards compatible, they implement a more complex `Behaviour` trait which combines `gossipsub` with `floodsub`.  There would be standard logic, such as subscribing to topic `X` should subscribe to the floodsub behaviour and the gossipsub behaviour. 

It would mean this implementation deviates from the spec and user's of rust-libp2p would have to manually implement the more complex behaviour (I could add this in an example, such that user's can copy it). 

What are your thoughts on this?

Reply:
> For example, sending a message to topic X, should send to all peers that implement the gossipsub protocol that are in mesh and also to all floodsub peers.

That looks wrong to me. Why would gossipsub exist if you have to mimic all the messages on floodsub as well?

However to back to the point, another possibility is to use the `FloodsubBehaviour` from within the `GossipsubBehaviour`.

Reply:
> For example, sending a message to topic X, should send to all peers that implement the gossipsub protocol that are in mesh and also to all floodsub peers.

See for example: https://github.com/libp2p/specs/tree/master/pubsub/gossipsub#message-processing

The idea is that gossipsub is backwards compatible. It knows of peers which use the floodsub protocol and behaves like another floodsub peer to those peers. For gossipsub peers it performs a more complex routing system. 

This could be my ignorance of rust-libp2p, but even if I had a `FloodsubBehaviour` as a field inside `GossipsubBehaviour` I'm still limited by the `NetworkBehaviour` trait which specifies the input to the behaviour (i.e. `inject_connected()` `inject_node_event()`). With these, I cannot distinguish which peer I should perform floodsub logic and which peer I should perform gossipsub logic. Am I mistaken in the understanding that this is the purpose of building complex `NetworkBehaviour`'s which is assisted via the `derive(NetworkBehaviour)` and implementing the `NetworkBehaviourEventProcess`?

Reply:
> which peer I should perform floodsub logic and which peer I should perform gossipsub logic.

That's highly dependant on the logic of gossipsub.
I suppose that you would assume that all peers support floodsub, and then you track through a hashmap which ones support gossipsub.

Reply:
Yes, I had this hashmap. The problem I have is how to populate such a hashmap. `inject_connected()` does not distinguish between nodes that support different protocols.

Reply:
That's where the protocol logic comes into play.
You can try to send a message and check whether the handler receives an "unsupported" error back, or you can add a method in the API so that the outer layer has to perform an identification.

Reply:
Yep. When a node connects, I'll try and identify it by seeing if the handler receives an "unsupported" error. 
I don't know about js-libp2p, but with go-libp2p, when the node connects you can determine which protocol it supports and simply add it to the floodsub peer or gossipsub peer mapping.